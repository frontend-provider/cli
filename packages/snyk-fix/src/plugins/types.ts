import { EntityToFix } from '../types';

export type FixHandler = (entities: EntityToFix[]) => Promise<FixHandlerResult[]>;

export interface FixHandlerResult {
  succeeded: EntityToFix[];
  failed: EntityToFix[];
}
