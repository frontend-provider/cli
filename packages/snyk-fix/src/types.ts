import { DepGraphData } from '@snyk/dep-graph';
import { CustomError } from './lib/errors/custom-error';

/* Scan Result
 * this data is returned by the CLI plugins to identify
 * what should be scanned for issues
 */
export interface GitTarget {
  remoteUrl: string;
  branch: string;
}
export interface ContainerTarget {
  image: string;
}

export interface ScanResult {
  readonly identity: Identity;
  readonly facts: Facts[];
  readonly name?: string;
  readonly policy?: string;
  readonly target?: GitTarget | ContainerTarget;
}

export interface Identity {
  type: string;
  targetFile?: string;
  // options used to scan should be here
  args?: { [key: string]: string };
}

export interface Facts {
  type: string;
  data: any;
}

/* Test Result
 * this data is returned on a `snyk test` for supported project types
 * after the relevant plugin extracts dependencies
 */
export interface TestResult {
  readonly issues: Issue[];
  readonly issuesData: IssuesData;
  readonly depGraphData: DepGraphData;
  readonly remediation?: RemediationChanges; // TODO: not yet in the CLI TestResults type
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

export interface FixInfo {
  upgradePaths?: UpgradePath[];
  isPatchable?: boolean;
  nearestFixedInVersion?: string;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}
export interface IssuesData {
  [issueId: string]: {
    id: string;
    severity: SEVERITY;
    title: string;
  };
}

/* Remediation Data
 * this data is returned on a `snyk test` for supported project types
 */
export interface Upgrade {
  upgradeTo: string; // name@version
}

export interface UpgradeVulns extends Upgrade {
  vulns: string[];
}
export interface UpgradeRemediation extends UpgradeVulns {
  upgrades: string[];
}

export interface PatchRemediation {
  paths: PatchObject[];
}

export interface DependencyUpdates {
  [from: string]: UpgradeRemediation;
}

export interface PinRemediation extends UpgradeVulns {
  isTransitive: boolean;
}

export interface DependencyPins {
  [name: string]: PinRemediation;
}

// Remediation changes to be applied to the project,
// including information on all and unresolved issues.
export interface RemediationChanges {
  unresolved: IssueData[];
  upgrade: DependencyUpdates;
  patch: {
    [name: string]: PatchRemediation;
  };
  ignore: unknown;
  pin: DependencyPins;
}

export interface IssueData {
  id: string;
  packageName: string;
  version: string;
  moduleName?: string;
  below: string; // Vulnerable below version
  semver: {
    vulnerable: string | string[];
    vulnerableHashes?: string[];
    vulnerableByDistro?: {
      [distroNameAndVersion: string]: string[];
    };
  };
  patches: Patch[];
  isNew: boolean;
  description: string;
  title: string;
  severity: SEVERITY;
  fixedIn: string[];
  legalInstructions?: string;
  reachability?: REACHABILITY;
}

interface Patch {
  version: string;
  id: string;
  urls: string[];
  modificationTime: string;
}

export enum REACHABILITY {
  FUNCTION = 'function',
  PACKAGE = 'package',
  NOT_REACHABLE = 'not-reachable',
  NO_INFO = 'no-info',
}

export interface PatchObject {
  [name: string]: {
    patched: string;
  };
}

export enum SEVERITY {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

/* End Remediation Data
 */

/* Snyk fix types
 * Types for concepts introduced as part of this lib
 */

export type SupportedScanTypes = 'pip';

export interface Workspace {
  path: string;
  readFile: (path: string) => Promise<string>;
  writeFile: (path: string, content: string) => Promise<void>;
}
export interface EntityToFix {
  readonly workspace: Workspace;
  readonly scanResult: ScanResult;
  readonly testResult: TestResult;
  readonly options: CliTestOptions;
}

// Partial CLI test options interface
// defining only what is used by @snyk/fix
// add more as needed
export interface PythonTestOptions {
  command?: string; // python interpreter to use for python tests
  dev?: boolean;
}
export type CliTestOptions = PythonTestOptions;
export interface WithError<Original> {
  original: Original;
  error: CustomError;
  tip?: string;
}

export interface WithFixChangesApplied<Original> {
  original: Original;
  changes: FixChangesSummary[];
}

export interface WithUserMessage<Original> {
  original: Original;
  userMessage: string;
}

export type FixChangesSummary = FixChangesSuccess | FixChangesError;

interface FixChangesSuccess {
  success: true;
  userMessage: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

interface FixChangesError {
  success: false;
  userMessage: string;
  reason: string;
  tip?: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

export interface ErrorsByEcoSystem {
  [ecosystem: string]: { originals: EntityToFix[]; userMessage: string };
}
export interface FixOptions {
  dryRun?: boolean;
  quiet?: boolean;
  stripAnsi?: boolean;
}

export interface FixedMeta {
  fixed: number;
  failed: number;
  fixableIssues: number;
  fixedIssues: number;
  totalIssues: number;
}
