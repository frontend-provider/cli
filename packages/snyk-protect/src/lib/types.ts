export type PatchInfo = {
  patchableVersions?: string;
  urls: string[];
};

export type VulnIdAndPackageName = {
  vulnId: string;
  packageName: string;
};

export type FoundPhysicalPackage = {
  packageName: string;
  packageVersion: string;
  path: string;
};

export type Patch = {
  patchableVersions?: string;
  patchDiffs: string[];
};

export type VulnPatches = {
  vulnId: string;
  patches: Patch[];
};

export type PatchedModule = {
  vulnId: string;
  packageName: string;
  packageVersion: string;
};

export enum ProtectResultType {
  NO_SNYK_FILE = 'NO_SNYK_FILE',
  NOTHING_TO_PATCH = 'NOTHING_TO_PATCH',
  APPLIED_PATCHES = 'APPLIED_PATCHES',
}

export type AnalyticsPayload = {
  command: string;
  args: string[];
  version: string;
  nodeVersion: string;
  metadata: {
    protectResult: ProtectResult;
  };
};

export type NoSnykFile = {
  type: ProtectResultType.NO_SNYK_FILE;
};

export type NothingToPatch = {
  type: ProtectResultType.NOTHING_TO_PATCH;
};

export type AppliedPatches = {
  type: ProtectResultType.APPLIED_PATCHES;
  patchedModules: PatchedModule[];
};

export type ProtectResult = NoSnykFile | NothingToPatch | AppliedPatches;
