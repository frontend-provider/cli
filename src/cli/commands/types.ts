export type MethodResult = CommandResult | string | void;

export class CommandResult {
  result: string;
  constructor(result: string) {
    this.result = result;
  }

  public toString(): string {
    return this.result;
  }

  public getDisplayResults() {
    return this.result;
  }
}

export abstract class TestCommandResult extends CommandResult {
  protected jsonResult = '';
  protected sarifResult = '';

  public getJsonResult(): string {
    return this.jsonResult;
  }

  public getSarifResult(): string {
    return this.sarifResult;
  }

  public static createHumanReadableTestCommandResult(
    humanReadableResult: string,
    jsonResult: string,
    sarifResult?: string,
  ): HumanReadableTestCommandResult {
    return new HumanReadableTestCommandResult(
      humanReadableResult,
      jsonResult,
      sarifResult,
    );
  }

  public static createJsonTestCommandResult(
    stdout: string,
    jsonResult?: string,
    sarifResult?: string,
  ): JsonTestCommandResult {
    return new JsonTestCommandResult(stdout, jsonResult, sarifResult);
  }
}

class HumanReadableTestCommandResult extends TestCommandResult {
  protected jsonResult = '';
  protected sarifResult = '';

  constructor(
    humanReadableResult: string,
    jsonResult: string,
    sarifResult?: string,
  ) {
    super(humanReadableResult);
    this.jsonResult = jsonResult;
    if (sarifResult) {
      this.sarifResult = sarifResult;
    }
  }

  public getJsonResult(): string {
    return this.jsonResult;
  }

  public getSarifResult(): string {
    return this.sarifResult;
  }
}

class JsonTestCommandResult extends TestCommandResult {
  constructor(stdout: string, jsonResult?: string, sarifResult?: string) {
    super(stdout);
    if (jsonResult) {
      this.jsonResult = jsonResult;
    }
    if (sarifResult) {
      this.sarifResult = sarifResult;
    } else {
      this.jsonResult = stdout;
    }
  }

  public getJsonResult(): string {
    return this.jsonResult;
  }

  public getSarifResult(): string {
    return this.sarifResult;
  }
}
