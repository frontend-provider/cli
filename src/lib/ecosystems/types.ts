import { DepGraphData } from '@snyk/dep-graph';
import { Options } from '../types';

export type Ecosystem = 'cpp' | 'docker';

export interface PluginResponse {
  scanResults: ScanResult[];
}

export interface GitTarget {
  remoteUrl: string;
  branch: string;
}

export interface ContainerTarget {
  image: string;
}

export interface ScanResult {
  identity: Identity;
  facts: Facts[];
  name?: string;
  policy?: string;
  target?: GitTarget | ContainerTarget;
}

export interface Identity {
  type: string;
  targetFile?: string;
  args?: { [key: string]: string };
}

export interface Facts {
  type: string;
  data: any;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

interface FixInfo {
  upgradePaths: UpgradePath[];
  isPatchable: boolean;
  nearestFixedInVersion?: string;
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

export interface IssuesData {
  [issueId: string]: {
    id: string;
    severity: string;
    title: string;
  };
}

export interface TestResult {
  issues: Issue[];
  issuesData: IssuesData;
  depGraphData: DepGraphData;
}

export interface EcosystemPlugin {
  scan: (options: Options) => Promise<PluginResponse>;
  display: (
    scanResults: ScanResult[],
    testResults: TestResult[],
    errors: string[],
    options: Options,
  ) => Promise<string>;
}
