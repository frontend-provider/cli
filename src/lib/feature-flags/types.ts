export type OrgFeatureFlagResponse = {
  ok?: boolean;
  userMessage?: string;
  code?: number;
  error?: string;
};
