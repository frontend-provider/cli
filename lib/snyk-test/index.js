module.exports = test;

var detect = require('../detect');
var runTest = require('./run-test');
var chalk = require('chalk');

function test(root, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  if (!options) {
    options = {};
  }

  var promise = executeTest(root, options);
  if (callback) {
    promise.then(function (res) {
      callback(null, res);
    }).catch(callback);
  }
  return promise;
}

function executeTest(root, options) {
  try {
    var packageManager = detect.detectPackageManager(root, options);
    options.packageManager = packageManager;
    return run(root, options)
      .then(function (res) {
        if (!res.packageManager) {
          res.packageManager = packageManager;
        }
        return res;
      });
  } catch (error) {
    return Promise.reject(chalk.red.bold(error));
  }
}

function run(root, options) {
  var packageManager = options.packageManager;
  if (['npm', 'yarn'].indexOf(packageManager) >= 0) {
    return require('./npm')(root, options);
  }
  if (!options.docker && [
    'rubygems',
    'maven',
    'gradle',
    'sbt',
    'pip',
    'golangdep',
    'govendor',
    'nuget',
    'composer',
  ].indexOf(packageManager) === -1) {
    throw new Error('Unsupported package manager: ' + packageManager);
  }
  return runTest(packageManager, root, options);
}
