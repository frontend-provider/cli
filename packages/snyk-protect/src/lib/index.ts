import * as fs from 'fs';
import * as path from 'path';
import { extractPatchMetadata } from './snyk-file';
import { applyPatchToFile } from './patch';
import { findPhysicalModules } from './explore-node-modules';
import {
  VulnIdAndPackageName,
  VulnPatches,
  PatchedModule,
  ProtectResultType,
} from './types';
import { getAllPatches } from './fetch-patches';
import { sendAnalytics } from './analytics';

async function protect(projectFolderPath: string) {
  const snykFilePath = path.resolve(projectFolderPath, '.snyk');

  if (!fs.existsSync(snykFilePath)) {
    console.log('No .snyk file found');
    sendAnalytics({
      type: ProtectResultType.NO_SNYK_FILE,
    });
    return;
  }

  const snykFileContents = fs.readFileSync(snykFilePath, 'utf8');
  const snykFilePatchMetadata = extractPatchMetadata(snykFileContents);
  const vulnIdAndPackageNames: VulnIdAndPackageName[] = snykFilePatchMetadata;
  const targetPackageNames = [
    ...new Set(snykFilePatchMetadata.map((vpn) => vpn.packageName)), // get a list of unique package names by converting to Set and then back to array
  ];

  // find instances of the target packages by spelunking through the node_modules looking for modules with a target packageName
  const foundPhysicalPackages = findPhysicalModules(
    projectFolderPath,
    targetPackageNames,
  );

  // Map of package name to versions (for the target package names).
  // For each package name, we might have found multiple versions and we'll need to fetch patches for each version.
  // We will use this to lookup the versions of packages to get patches for.
  const packageNameToVersionsMap = new Map<string, string[]>();
  foundPhysicalPackages.forEach((p) => {
    if (packageNameToVersionsMap.has(p.packageName)) {
      const versions = packageNameToVersionsMap.get(p.packageName);
      if (!versions?.includes(p.packageVersion)) {
        versions?.push(p.packageVersion);
      }
    } else {
      packageNameToVersionsMap.set(p.packageName, [p.packageVersion]);
    }
  });

  const packageAtVersionsToPatches: Map<
    string,
    VulnPatches[]
  > = await getAllPatches(vulnIdAndPackageNames, packageNameToVersionsMap);

  if (packageAtVersionsToPatches.size === 0) {
    console.log('Nothing to patch');
    sendAnalytics({
      type: ProtectResultType.NOTHING_TO_PATCH,
    });
    return;
  }

  const patchedModules: PatchedModule[] = [];
  foundPhysicalPackages.forEach((fpp) => {
    const packageNameAtVersion = `${fpp.packageName}@${fpp.packageVersion}`;
    const vuldIdAndPatches = packageAtVersionsToPatches.get(
      packageNameAtVersion,
    );
    vuldIdAndPatches?.forEach((vp) => {
      vp.patches.forEach((patchDiffs) => {
        patchDiffs.patchDiffs.forEach((diff) => {
          applyPatchToFile(diff, fpp.path);
        });
      });
      patchedModules.push({
        vulnId: vp.vulnId,
        packageName: fpp.packageName,
        packageVersion: fpp.packageVersion,
      });
    });
  });

  console.log('Successfully applied Snyk patches');

  sendAnalytics({
    type: ProtectResultType.APPLIED_PATCHES,
    patchedModules,
  });
}

export default protect;
