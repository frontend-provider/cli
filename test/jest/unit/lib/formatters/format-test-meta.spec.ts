import stripAnsi from 'strip-ansi';

import { formatTestMeta } from '../../../../../src/lib/formatters/format-test-meta';
import { IacTestResponse } from '../../../../../src/lib/snyk-test/iac-test-result';
import { TestResult } from '../../../../../src/lib/snyk-test/legacy';
import { ShowVulnPaths } from '../../../../../src/lib/types';
import { generateCloudConfigResults } from '../../iac-unit-tests/results-formatter.fixtures';

describe('formatTestMeta', () => {
  it('iacTestResult meta is formatted as expected', () => {
    const iacTestResult: IacTestResponse = {
      path: 'src',
      targetFile: 'example.yaml',
      projectName: 'snyk',
      displayTargetFile: 'src/example.yaml',
      foundProjectCount: 0,
      meta: {
        isLicensesEnabled: false,
        org: 'org-name',
        isPublic: false,
        policy: '',
      },
      result: {
        cloudConfigResults: generateCloudConfigResults(false),
        projectType: 'k8sconfig',
      },
      ok: true,
      isPrivate: false,
      org: 'test-org',
      summary: 'No known vulnerabilities',
    };
    const options = {
      iac: true,
      path: '/path/to/test',
      showVulnPaths: 'all' as ShowVulnPaths,
    };
    expect(stripAnsi(formatTestMeta(iacTestResult, options))).toMatchSnapshot();
  });
  it('with TargetFile', () => {
    const testResult: TestResult = {
      targetFile: 'package.json',
      packageManager: 'npm',
      vulnerabilities: [],
      dependencyCount: 8,
      policy: '',
      licensesPolicy: null,
      ignoreSettings: null,
      ok: true,
      org: 'my-org',
      isPrivate: false,
      summary: 'No known vulnerabilities',
    };
    const options = {
      path: '/path/to/test',
      showVulnPaths: 'all' as ShowVulnPaths,
    };
    expect(stripAnsi(formatTestMeta(testResult, options))).toMatchSnapshot();
  });

  it('without TargetFile & with license Policy', () => {
    const testResult: TestResult = {
      packageManager: 'pip',
      vulnerabilities: [],
      dependencyCount: 8,
      policy: '',
      licensesPolicy: {
        severities: {},
        orgLicenseRules: {
          'AGPL-1.0': {
            licenseType: 'AGPL-1.0',
            severity: 'high',
            instructions: '',
          },
        },
      },
      ignoreSettings: null,
      ok: true,
      org: 'my-org',
      isPrivate: false,
      summary: 'No known vulnerabilities',
    };
    const options = {
      path: '/path/to/test',
      showVulnPaths: 'all' as ShowVulnPaths,
    };
    expect(stripAnsi(formatTestMeta(testResult, options))).toMatchSnapshot();
  });
  it('Docker', () => {
    const testResult: TestResult = {
      vulnerabilities: [],
      dependencyCount: 8,
      packageManager: 'deb' as any, // this is returned from the backend after test
      platform: 'linux/amd64',
      policy: '',
      docker: {
        baseImage: 'debian:latest',
      },
      licensesPolicy: null,
      ignoreSettings: null,
      ok: true,
      org: 'my-org',
      isPrivate: false,
      summary: 'No known vulnerabilities',
    };
    const options = {
      path: 'my-image:latest',
      docker: true,
      showVulnPaths: 'all' as ShowVulnPaths,
    };
    expect(stripAnsi(formatTestMeta(testResult, options))).toMatchSnapshot();
  });
  it('with filesystem policy', () => {
    const testResult: TestResult = {
      packageManager: 'pip',
      vulnerabilities: [],
      dependencyCount: 8,
      filesystemPolicy: true,
      policy:
        "# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.\nversion: v1.13.1\nignore: {}\n# patches apply the minimum changes required to fix a vulnerability\npatch:\n  'npm:qs:20170213':\n    - npm-package-with-git-url > qs:\n        patched: '2018-11-04T12:47:13.696Z'\n",
      licensesPolicy: {
        severities: {},
        orgLicenseRules: {
          'AGPL-1.0': {
            licenseType: 'AGPL-1.0',
            severity: 'high',
            instructions: '',
          },
        },
      },
      ignoreSettings: null,
      ok: true,
      org: 'my-org',
      isPrivate: false,
      summary: 'No known vulnerabilities',
    };
    const options = {
      path: '/path/to/test',
      showVulnPaths: 'all' as ShowVulnPaths,
    };
    expect(stripAnsi(formatTestMeta(testResult, options))).toMatchSnapshot();
  });
});
