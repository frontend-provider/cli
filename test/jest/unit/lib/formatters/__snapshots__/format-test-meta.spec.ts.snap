// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatTestMeta Docker 1`] = `
"Organization:      my-org
Package manager:   deb
Docker image:      my-image:latest
Platform:          linux/amd64
Base image:        debian:latest"
`;

exports[`formatTestMeta iacTestResult meta is formatted as expected 1`] = `
"Organization:      test-org
Type:              Infrastructure as Code
Target file:       example.yaml
Project name:      snyk
Open source:       yes
Project path:      /path/to/test"
`;

exports[`formatTestMeta with TargetFile 1`] = `
"Organization:      my-org
Package manager:   npm
Target file:       package.json
Open source:       yes
Project path:      /path/to/test"
`;

exports[`formatTestMeta with filesystem policy 1`] = `
"Organization:      my-org
Package manager:   pip
Open source:       yes
Project path:      /path/to/test
Local Snyk policy: found
Licenses:          enabled"
`;

exports[`formatTestMeta without TargetFile & with license Policy 1`] = `
"Organization:      my-org
Package manager:   pip
Open source:       yes
Project path:      /path/to/test
Licenses:          enabled"
`;
